File: backend/app/models/friend.py
from datetime import datetime, timezone
from typing import Optional, Dict, Any
from sqlmodel import SQLModel, Field
from sqlalchemy import UniqueConstraint, Column, Enum as SAEnum, DateTime, text
from enum import Enum


def utc_now() -> datetime:
    """Return current UTC time as timezone-aware datetime."""
    return datetime.now(timezone.utc)


class Friend(SQLModel, table=True):
    __tablename__ = "friends"
    __table_args__ = (
        UniqueConstraint("user_id", "friend_id", name="uix_user_friend"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id", ondelete="CASCADE", index=True)
    friend_id: int = Field(foreign_key="users.id", ondelete="CASCADE", index=True)
    created_at: datetime = Field(
        default_factory=utc_now,
        sa_column=Column(DateTime(timezone=True), server_default=text("now()")),
    )


class StatusRequestType(str, Enum):
    PENDING = "PENDING"
    ACCEPTED = "ACCEPTED"
    REJECTED = "REJECTED"


class FriendRequest(SQLModel, table=True):
    __tablename__ = "friend_requests"
    __table_args__ = (
        UniqueConstraint("sender_id", "receiver_id", name="uix_sender_receiver"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    sender_id: int = Field(foreign_key="users.id", ondelete="CASCADE", index=True)
    receiver_id: int = Field(foreign_key="users.id", ondelete="CASCADE", index=True)
    status: StatusRequestType = Field(
        sa_column=Column(SAEnum(StatusRequestType, name="status_request_type_enum")),
        default=StatusRequestType.PENDING,
    )
    created_at: datetime = Field(
        default_factory=utc_now,
        sa_column=Column(DateTime(timezone=True), server_default=text("now()")),
    )
    responded_at: Optional[datetime] = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=True),
    )

File: frontend/src/types/friend.ts
export enum RelationshipStatus {
    FRIENDS = "FRIENDS",
    REQUEST_SENT = "REQUEST_SENT",
    REQUEST_RECEIVED = "REQUEST_RECEIVED",
    NONE = "NONE",
}

export enum StatusRequestType {
    PENDING = "PENDING",
    ACCEPTED = "ACCEPTED",
    REJECTED = "REJECTED",
}

export interface Friend {
    user_id: number;
    username: string;
    xp: number;
    streak: number;
};

export interface UserSearchResult {
    user_id: number;
    username: string;
    relationship_status: RelationshipStatus;
}

export interface FriendRequest {
    request_id: number;
    sender_id: number;
    sender_username: string;
    created_at: string;
}

export interface FriendRequestCreate {
    receiver_id: number;
}

export interface FriendRequestReturn extends FriendRequest {
    id: number;
    sender_id: number;
    receiver_id: number;
    status: StatusRequestType;
    created_at: string;
}